compiler/rustc_errors/src/emitter.rs

```diff
@@ -1678,6 +1678,23 @@ impl EmitterWriter {
 
                 if is_multiline {
                     for SubstitutionHighlight { start, end } in parts {
+                        // Account for tabs when highlighting (#87972).
+                        let start: usize = line
+                            .chars()
+                            .take(*start)
+                            .map(|ch| match ch {
+                                '\t' => 4,
+                                _ => unicode_width::UnicodeWidthChar::width(ch).unwrap_or(1),
```

Comment:
Can you verify the current version of the PR against your test cases? Sadly, we don't currently have regression tests for the coloring of the output in our tests :-/

Reply:
Yup, the highlighting looks good now!

<img width="830" alt="Screenshot 2021-08-12 at 21 38 57" src="https://user-images.githubusercontent.com/783247/129258461-5e97f40f-1410-4708-b7fc-cf7386a2ce53.png">

Reply:
In particular, I want to confirm that [the output of the terminal without emoji support you showed here](https://github.com/rust-lang/rust/issues/87972#issuecomment-897767010) is reasonable (as I fully expect it to be, because it relies on what we were doing for the multiline case, which looks ok in that screenshot).

Reply:
Sadly, now I'm seeing _other_ issues :(